Dependency tools

hiker opened the pull request:
This pull requests adds a class with some handy tools based on the dependency analysis. I am now using this tool for my ROMS work, and it removes over 100 lines from my ROMS script.

The one thing I am not sure of: where do you want to put this class? For now I have put it in psyclone/src/psyclone, but it could go in psyclone/src/psyclone/tools (new directory)?  Or maybe psyclone/tools/? It could also go to psyclone/examples ... but imho it is a bit more than an example (and an example should not be documented in the manual, while I have added documentation to the developers guide).

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6160,20 +6272,41 @@ def reference_accesses(self, var_accesses):
 
         # Merge the data (that shows now WRITE for the variable) with the
-        # parameter to this function:
+        # parameter to this function. It is important that first the
```
Good explanation - thanks.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6224,6 +6357,15 @@ def view(self, indent=0):
+    def math_equal(self, other):
+        '''Returns true if the self has the same results as other.
+        :param other: the node to compare self with.
+        :type other: py:class:`psyclone.psyGen.Node`
```
Please document the :returns: and :rtype:

hiker replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6224,6 +6357,15 @@ def view(self, indent=0):
         return "Reference[name:'" + self._reference + "']"
 
+    def math_equal(self, other):
+        '''Returns true if the self has the same results as other.
```
s/true/True/

hiker replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6374,6 +6516,29 @@ class BinaryOperation(Operation):
         ])
 
+    def math_equal(self, other):
+        '''Returns true if the self has the same results as other.
```
s/true/True/

arporter replied:
and please document the return value.

hiker replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6529,6 +6694,15 @@ def view(self, indent=0):
         return "Literal[value:'" + self._value + "']"
 
+    def math_equal(self, other):
+        '''Returns true if the self has the same results as other.
```
Please replace the summary text with docstrings for :returns: and :rtype:

hiker replied:
Done.

arporter commented:
I forgot to say that I can see these new tools being very useful for NEMO too (which is the point really). As they're part of the PSyclone package they should probably go in `src/psyclone/tools` but we should consult @rupertford about this as well.

hiker commented on doc/developer_guide/developers.rst:
```diff
@@ -818,6 +818,43 @@ until we find accesses that would prevent parallelisation::
+          var = statement.variable_name
+          # Check if there is a variable dependency that might 
+          # prevent this loop from being parallelised:
+          if dt.can_loop_be_parallelised(statement, var):
```
Good point. Originally I thought to keep it general enough to put any statements in, but then I didn't need it. But I still like to have that option (and maybe in the future we could remove the condition that the node is a loop). As a compromise I have made it optional: if it's not specified, I'll use the loop variable. If you prefer to remove it, let me know.

hiker replied:
OK, I am checking what I have missed. I changed the code, but not the example ... done.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -74,6 +74,12 @@ def __init__(self, access_type, location, node, indices=None):
             self._indices = None
 
+    def __str__(self):
+        ''':returns: a string representating showing the access mode
```
Done, and added to the coding style guide.

arporter replied:
Thanks.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -136,15 +142,32 @@ class VariableAccessInfo(object):
         self._accesses = []
 
+    def __str__(self):
+        ''':returns: a string representation of this object with the format:
```
Done

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -136,15 +142,32 @@ class VariableAccessInfo(object):
         '''
         return self._var_name
 
+    def is_array(self):
```
Good question. Ideally, I would be using the symbol table to find out if the variable is an array. Then I could use is_indexed_array_access as a test for each access (e.g. to detect array assignments - though there might be other methods to do that??). But at this stage is_array tests if all accesses are indexed array accesses, so I think the name is_indexed_array_access(es) is not that good.

This function should tell if the variable is an array - till we have a symbol table everywhere, I propose to leave the function with this name (am happy to add a #TODO - we can discuss this at the next telco).

hiker replied:
Done.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -136,15 +142,32 @@ class VariableAccessInfo(object):
+    def is_array(self):
+        ''':returns: true if all accesses to this variable involve indices.
+        :rtype: bool'''
+        for access in self._accesses:
```
Indeed, good point. And not only here. I've rewritten all those functions to use various all/any statements, which I think makes it much easier to understand them (plus it should be faster).

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -157,6 +180,18 @@ def is_written(self):
         return False
 
+    def is_read_only(self):
+        '''Checks if the specified variable name is always read, and never
```
Done

hiker replied:
Oh, I missed quite a few of those :(

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -328,10 +363,13 @@ def add_access(self, var_name, access_type, node, indices=None):
-        ''':returns: all variables contained in this instance.
+        ''':returns: all variables contained in this instance, sorted (in \
+                     order to make test results reproducible).
         :rtype: List of str.
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+    a messaging system where functions can store messages that might be
+    useful for the user to see.
+
+    :param loop_types_to_parallelise: A list of loop types that will be \
```
Documented.

hiker replied:
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+                                      considered for parallelisation. This \
+                                      can be used to exclude say 1-dimensional\
+                                      loops.
+    :type  loop_types_to_parallelise: list of str
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+
+    # -------------------------------------------------------------------------
+    def _add_info(self, message):
+        '''Adds a informational message to the internal message
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        :type var_info: \
+            :py:class:`psyclone.core.access_info.VariableAccessInfo`
+
+        :return: if the variable can be used in parallel.
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+                # TODO #363: derived types are not supported, which
+                # atm have an index of type str
+                if isinstance(index, str):
+                    visitor = FortranWriter()
```
Oops - indeed. Gone.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+            list_of_indices = access.indices
+            # Now determine all dimensions that depend on the
+            # parallel variable:
+            for i, index in enumerate(list_of_indices):
```
Yes, it does ... but I have renamed all variables (including i).

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+            list_of_indices = access.indices
+            # Now determine all dimensions that depend on the
+            # parallel variable:
+            for i, index in enumerate(list_of_indices):
```
I renamed this to 'index_expression', since it will e.g. contain 'i+1-n' or whatever :)

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        # indices that are actually used
+        dimension_index = -1
+        all_indices = []
+        for access in var_info.all_accesses:
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+
+        # Array variable does not depend on parallel loop variable:
+        if not all_indices:
+            # Variable accesses an array that is not dependend on the parallel'
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+            #    a(1) = b(j)+1
+            #    c(j) = a(1) * 2
+            # enddo
+            # Either the variable should be declard to be a scalar, or
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+            #    c(j) = a(1) * 2
+            # enddo
+            # Either the variable should be declard to be a scalar, or
+            # it is a shared variable with constant access, which is not
```
Reworded.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+                                 variables_to_ignore=None,
+                                 var_accesses=None):
+        # pylint: disable=too-many-arguments
+        '''This function analyses the loop in PsyIR to see if
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        it can be safely parallelised over the specified variable.
+
+        :param loop: the loop node to be analysed.
+        :type loop: :py:class:`psyclone.psyGen.Node`
```
Indeed.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        :param loop: the loop node to be analysed.
+        :type loop: :py:class:`psyclone.psyGen.Node`
+        :param str loop_variable: name of the variable that is parallelised.
+        :param bool only_nested_loops: if true, a loop must have an inner \
```
I've changed the case of a whole set of trues to Trues :)

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        :param bool only_nested_loops: if true, a loop must have an inner \
+                                       loop in order to be considered \
+                                       parallelisable (default: True).
+        :param bool test_all_variables: if true, it will test if all variables\
```
Yes

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+                                        can be parallelised, otherwise it will\
+                                        stop after the first variable is found\
+                                        that can not be parallelised.
+        :param variables_to_ignore: List of variables that are not checked if\
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        :type var_accesses: \
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        :returns: true if the loop can be parallelised.
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        :returns: true if the loop can be parallelised.
+        :rtype: bool.
```
Done

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+        self._clear_messages()
+
+        if not isinstance(loop, Loop):
+            raise InternalError("can_loop_be_parallelised: Loop must be an "
```
Agree. TypeError is better.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -969,6 +970,27 @@ def __init__(self, ast=None, children=None, parent=None):
+            return False
+
+        for i, entity in enumerate(self.children):
+            if not entity.math_equal(other.children[i]):
```
I did notice your comment, but I still need to clarify this: this code is very limited in its scope. E.g. the expressions 'i+n+1' and 'i+1+n' would likely not be considered equal (assuming that '+' binds left first: (i+n)+1 and (i+1)+n. We could of course extend it, but the question is how much effort do we want to spend in this? At a certain stage it might be simpler to plug in an (optional??) symbolic maths package - afaik something like this exists for python.

Should we perhaps even remove this option for now, considering that it is limited? Or extend it to work at least for 'n' identical binary operations (e.g. i+n+1 as above)? The moment we support more than one operation (i+5*k+1 vs i+1+k*5 ) ... it becomes a lot more complicated.

hiker replied:
OK, I've created #533, documented the current behaviour in the developer guide, added more test cases that show what works, and an xfailing one with a reference to #533 showing what does not work.

arporter replied:
You're a star, thanks.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -2542,9 +2564,26 @@ def update(self):
-        # If end_idx+1 takes us beyond the range of the list then the
-        # element is appended to the list
-        self._parent.ast.content.insert(end_idx+1, enddir)
+        # FIXME: In case of:
```
Yes, from memory it's all related to the AST. Since (afaik) this is only temporary till the PSyIR is fully supported  - tb(very)h I don't care, I would leave it in till the whole update functions can be removed (add a #todo to a psyir ticket ).

hiker replied:
I believe that should be #435, which I have added (#500 is use SymbolTable).

arporter replied:
Oh yes, thanks.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -2542,9 +2564,26 @@ def update(self):
+        # In order to avoid this problem when an "omp do" is present, test
+        # for this case and if so move the "omp end parallel" two statements
+        # further down, i.e. after the loop and "omp end do" statement.
+        if isinstance(self._parent.ast.content[end_idx], Comment) and \
```
My gut feeling when I looked at the code again was that we need to call the update of the child after adding the start, before adding the end. But same reason as above: this is imho temporary till psyir works (and iirc with my directive support this part should already work, and I am happy to add openacc directive support as well), and I am happy to just leave it.

hiker replied:
I need an update here - can we leave this till #435 is done (which I have added a TODO for).

arporter replied:
OK, since this is just an additional fix on top of what was there already and you've added a TODO I'm just about OK with leaving it as is.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -2662,6 +2701,79 @@ def gen_code(self, parent):
+        # This initial loop is necessary in case that the parent is e.g. an
+        # omp parallel node (which has no ast either). So we search up
+        # till we found the first tree to actually have an ast.
+        while not parent.ast:
```
And again ... this seems to be temporary.

hiker replied:
#435, done.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -2662,6 +2701,79 @@ def gen_code(self, parent):
+        # the AST representing our last child
+        enddir = Comment(FortranStringReader("!$omp end do",
+                                             ignore_comments=False))
+        if start_idx == len(parent.content) - 1:
```
Ah, thanks!

rupertford commented:
> I forgot to say that I can see these new tools being very useful for NEMO too (which is the point really). As they're part of the PSyclone package they should probably go in `src/psyclone/tools` but we should consult @rupertford about this as well.

If the tools are generic (useful for all APIs) then I would suggest that they be in a subdirectory of the PSyIR as this is what they operate on. I would extend this argument to the underlying dependency analysis tools, tree traversal and generic transformations i.e. anything that operates on the PSyIR in a generic way. Perhaps we could discuss this on Wednesday? I am sure @sergisiso would also have a view on this.

hiker commented:
I have to admit I absolutely [rage-quited](https://external-preview.redd.it/gL5bkwrl5Asm_8ICpBWJlnbe72kTiNcTHRP2f44gQhE.gif?format=mp4&s=1058f20982d0941b4c982e02cd6a258d6079e00d) this pull request, trying to cover one exception in a test. The logic was simple, I needed to set the parent in the AST to None:
```
omp_directive.parent.ast.parent = None
```
It took me hours to find out why this did not work, and it is probably indeed obvious to you. I really, really REALLY hate 😡 🔥  property-setters - with proper setter functions I would have known immediately ('directive.parent.ast.set_parent(None)' ... well, that would of course then be: 'directive.get_parent().get_ast().set_parent(None)' - I admit that's not nice 😀 ).

Of course, now I feel stupid (but my debug prints of ....ast.parent showed me it's None, just the test of None did not trigger 😳 ) - gee, such a waste of time. Anyway, I have that fixed now.

Unfortunately I also had to fix a bug in the handling of OMPParallel that then got triggered with my test cases. Those update() functions are a major PITA. I really hope that my understanding is correct that they will be removed once we can fully use the PSyIR? Can't wait for that (and imho the update functions for openmp are still wrong).


I think it's ready for the next review.

hiker commented:
> If the tools are generic (useful for all APIs) then I would suggest that they be in a subdirectory of the PSyIR as this is what they operate on. I would extend this argument to the underlying dependency analysis tools, tree traversal and generic transformations i.e. anything that operates on the PSyIR in a generic way. Perhaps we could discuss this on Wednesday? I am sure @sergisiso would also have a view on this.

I don't mind one way or another - but I might point out that I see a (tiny) difference between say the dependency collection stuff, and this tool, though I find it hard to explain: the dependency collection stuff is 'IN' psyclone, there is no way that a user could implement this functionality (without extending PSyclone). While this dependency tool sits (more or less) on top of python, any user could easily write and extend this tool - it feels closer to a user script, something to be used by a script, and not so much PSyclone itself. So from that point of view I would perhaps suggest a more 'visible' directory. but of course, if it is used by e.g. the NEMO api, it is kind of internal ... I am not too fussed tbh, just wanted to point this out.

arporter commented:
Hi @hiker, I've just come to look at this again and am not seeing some of the changes that you say you've made. Have you pushed all your changes?

hiker commented:
I indeed missed a few things, which I only applied partially - my apologies, and hopefully fixed now.

arporter commented on src/psyclone/tests/dependency_tools_test.py:
```diff
@@ -0,0 +1,289 @@
+    # Write only scalar variable: a(ji, jj) = b
+    parallel = dep_tools.can_loop_be_parallelised(loops[1], "jj")
+    assert not parallel
+    assert "Variable 'b' is only written once" \
```
Might be better to include "Scalar" in this message too.

hiker replied:
Good idea.

arporter commented on src/psyclone/tests/dependency_tools_test.py:
```diff
@@ -0,0 +1,289 @@
+def test_derived_type(parser):
+    '''Tests assignment to derived type variables. For now (see #363)
+    they are only partially supported, and as default assume that these
+    kind of statements are parallelisable.
```
Should that say "are not parallelisable"?

hiker replied:
Yes (I think I changed my mind long after having written the docs - originally I excepted derived types to be parallelisable, then I added the ability to explicitly disable this for given variables, which is much nicer, and changed the default, without changing the docs).

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2542,14 +2574,31 @@ def update(self):
+                "omp do" in str(parent_ast.content[end_idx]):
+            # We need to test for instance, otherwise the string representation
+            # of a loop could somewhere contain an "omp do"
+            parent_ast.content.insert(end_idx+3, enddir)
```
Please could you add a check here that `end_idx+3` really is the right location - i.e. that the nodes either side are what you expect. That way, if this 'fix' fails then at least we'll exit cleanly.

hiker replied:
Hmm - I did add one more test, but tbh there is only so much we can test. The test for end_idx+2 only confirms that an omp-do has exactly one inner node (the loop), and one end-do. The omp do transform should already guarantee that.

arporter commented on src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:
```diff
@@ -147,9 +147,80 @@ def test_omp_parallel_multi():
+    directive = new_sched.children[0].loop_body[1].else_body[0].else_body[0]\
+        .children[0]
+    assert isinstance(directive, OMPDoDirective)
+    directive.update()
```
Shouldn't this second call to `update()` be done before the check on the structure of the PSyIR? As it stands, the test immediately wipes the existing AST after this call to `update().`

hiker replied:
Yes. that code was somehow messed up ... and didn't test anything. I added a test that the old and new AST are actually the same.

arporter commented on src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:
```diff
@@ -147,9 +147,80 @@ def test_omp_parallel_multi():
+    psy = PSyFactory(API, distributed_memory=False).create(invoke_info)
+    schedule = psy.invokes.get('imperfect_nest').schedule
+    loop_trans = OMPLoopTrans()
+    new_sched, _ = loop_trans.apply(schedule
```
s/schedule.children[0]/schedule[0]/ throughout.

hiker replied:
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,348 @@
+                    self._add_warning("Assignment to derived type '{0}' is "
+                                      "not supported yet.".format(var_string))
+                    return False
+                index.reference_accesses(accesses)
```
I think that should be ok.

hiker commented on doc/developer_guide/developers.rst:
```diff
@@ -818,6 +818,43 @@ until we find accesses that would prevent parallelisation::
+.. autoclass:: psyclone.dependency_tools.DependencyTools
+    :members:
+
+Here an example on how to use this class. It takes a list of statements
```
Done.

hiker commented on doc/developer_guide/developers.rst:
```diff
@@ -818,6 +818,43 @@ until we find accesses that would prevent parallelisation::
+  parallel_loop = OMPLoopTrans()
+  # The loops in the Fortran functions that must be parallelised
+  # are over the 'grid' domain. Note that the psyclone config
+  # files specifies the mapping of loop variable to type, e.g.:
```
Done.

hiker commented on doc/developer_guide/developers.rst:
```diff
@@ -818,6 +818,43 @@ until we find accesses that would prevent parallelisation::
+  for statement in statements:
+      if isinstance(statement, NemoLoop):
+          # Get the loop variable of the potential parallel loop:
+          var = statement.variable_name
```
Indeed.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -136,49 +141,65 @@ class VariableAccessInfo(object):
+        '''Checks if this variable is always read, and never
+        written.
+
+        :returns: True if this variable is read only.
```
OK.

hiker replied:
I can't answer the previous comment about #500 for is_array ... which is done.

arporter replied:
and I can't resolve it either! Consider it resolved :-)

hiker commented on src/psyclone/core/tests/access_info_test.py:
```diff
@@ -124,20 +135,32 @@ def test_variable_access_info_read_write():
+    vai_array.add_access(AccessType.WRITE, 3, Node(), ["i"])
+    assert vai_array.is_array()
+
+    # Adding a non-array access marks this as not array:
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+    # -------------------------------------------------------------------------
+    def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
+        '''Simple first test to see if a loop should even be considered to
+        be parallelised. The default implementation test if the loop
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+                    return False
+                index_expression.reference_accesses(accesses)
+                try:
+                    _ = accesses[loop_variable]
```
Agreed with the first one, then I was going to write that 'in' can not be used, unless we implement __contain__ .... but then I realised that I can refactor the whole class to be a dict ... and then it all works rather nicely.

arporter replied:
Yay :-)

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+                try:
+                    _ = accesses[loop_variable]
+                    # This array would be parallelised across different
+                    # indices (e.g. a(i,j), and a(j,i) ):
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+                    # current dimension --> keep on looking
+                    continue
+
+        # The variable is an array variable, which does not depend
```
OK.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+        # The variable is an array variable, which does not depend
+        # on the loop variable at all (see examples below)
+        if not all_indices:
+            # An array is used that is not actually dependend on the parallel
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+        for index in all_indices[1:]:
+            if not first_index.math_equal(index):
+                visitor = FortranWriter()
+                self._add_warning("Variable {0} is using index {1} and {2} "
```
Done.

hiker commented on src/psyclone/dependency_tools.py:
```diff
@@ -0,0 +1,378 @@
+        # been tested above, so it must be a write access here, which prohibits
+        # parallelisation.
+        # We could potentially use lastprivate here?
+        if len(all_accesses) == 1:
```
Done.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -969,6 +970,30 @@ def __init__(self, ast=None, children=None, parent=None):
+        :param other: the node to compare self with.
+        :type other: py:class:`psyclone.psyGen.Node`.
+
+        :returns: if self has the same result as other.
```
Done. One day I will develop a natural sense of when to use if and when whether 😊

arporter replied:
I've only developed it in the last couple of years ;-)

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -1523,7 +1548,7 @@ def reductions(self, reprod=None):
 
     def is_openmp_parallel(self):
-        '''Returns true if this Node is within an OpenMP parallel region.
+        '''Returns True if this Node is within an OpenMP parallel region.
```
Done.

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -2521,14 +2546,21 @@ def update(self):
+        while parent_ast:
+            if hasattr(parent_ast, "content") and \
+                            parent_ast is not self.children[0].ast:
+                break
```
Oh I can't agree more ... error prone is an understatement. Code added, and #435 added (and also added a test for that).

hiker commented on src/psyclone/tests/dependency_tools_test.py:
```diff
@@ -0,0 +1,289 @@
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+
+''' Module containing tests the dependency tools.'''
```
Done.

hiker commented on src/psyclone/tests/dependency_tools_test.py:
```diff
@@ -0,0 +1,289 @@
+
+# -----------------------------------------------------------------------------
+def test_messages():
+    '''Tests the messaging system of the dependenc tools.'''
```
Done.

hiker commented on src/psyclone/tests/dependency_tools_test.py:
```diff
@@ -0,0 +1,289 @@
+    assert parallel
+    assert not dep_tools.get_all_messages()
+    # Make sure can_loop_be_parallelised clears old messages automatically
+    assert not dep_tools.get_all_messages()
```
Indeed not.

hiker commented on src/psyclone/tests/dependency_tools_test.py:
```diff
@@ -0,0 +1,289 @@
+    # Use a stencil access (with write), which prevents parallelisation
+    parallel = dep_tools.can_loop_be_parallelised(loops[3], "jj")
+    assert not parallel
+    assert "Variable mask is using index jj + 1 and jj and can therefore "\
```
Done.

hiker commented:
OK, I think I have addressed all comments, and moved the dependency_tools (and the test) into the corresponding subdirectories. 

The one thing I am not sure of is if we shouldn't move at least the dependency tool documentation into the user's guide? Maybe even the Var Access stuff as well? But I would actually say to leave this as a separate ticket for later.

Assuming that travis is happy: ready for next review.

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -0,0 +1,372 @@
+        self._clear_messages()
+
+        if not isinstance(loop, Loop):
+            raise TypeError("can_loop_be_parallelised: Loop must be an "
```
I've taken the liberty of doing this myself as I was just on the point of merging this PR when I realised I'd made this comment.

arporter commented:
I messed-up slightly and overlooked a (small) change that I was going to request. Since this only affects a docstring and the details of an error message I've done it myself. I'll merge once Travis has completed...
